hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+
+### The current replication model
+
+The current replication model in SCM/Ozone is very simplistic. We compute the replication count or the number of replications that we need to do as:
+
+```
+Replica count = expectedCount - currentCount
+```
+
+In case the _Replica count_ is positive, it means that we need to make more replicas. If the number is negative, it means that we are over replicated and we need to remove some replicas of this container. If the Replica count for a container is zero; it means that we have the expected number of containers in the cluster.
+
+To support idempontent placement strategies we include the in-fligt replications in the `currentCount`: If there are one in-flight replication process and two replicas we won't start a new replication command unless the original command is timed out.
+
+The timeout is configured with `hdds.scm.replication.event.timeout` and the default value is 10 minutes.
+
+More preciously the current algorithm is the following:
+
+```
+Replica count = expectedCount - healthy - inFlight
+```
+
+### The proposed solution
+
+To support the notion that a user can provide hints to the replication model, we propose to add two variables to the current model.
+
+In the new model, we propose to break the `currentCount` into the three separate groups. That is _Healthy nodes_, _Maintenance nodes_, and _Decommission nodes_. The new model replaces the currentCount with these three separate counts. The following function captures the code that drives the logic of computing Replica counts in the new model. The table below discusses the input and output of this model very extensively.
+
+```java
+/**
+ * Calculate the number of the missing replicas.
+ * 
+ * @return the number of the missing replicas. If it's less than zero, the container is over replicated.
+ */
+int getReplicationCount(int expectedCount, int healthy, 
+   int maintenance, int inFlight) {
+
+   //for over replication, count only with the healthy replicas
+   if (expectedCount < healthy) {
+      return expectedCount - healthy;
+   }
+   
+   replicaCount = expectedCount - (healthy + maintenance + inFlight);
+
+   if (replicaCount == 0 && healthy < 1) {
+      replicaCount ++;
+   }
+   
+   //over replication is already handled
+   return Math.max(0, replicaCount);
+}
+
+```
+
+We also need to specify two end condition when the DECOMMISSIONING node can be moved to the DECOMMISSIONED state or the ENTERING_MAINTENANCE mode can be moved to the IN_MAINTENANCE state.
+
+The following conditions should be true for all the containers and all the containers on the specific node should be closed.
+
+From DECOMMISSIONING to DECOMMISSIONED:
+
+ * There are at least one healthy replica
+ * There are at most one missing replica
+
+Which means that node can be decommissioned if: 
+ 
+ * all the containers with replication factor THREE have at least *one replica* on a HEALTHY nodes (minimum.live.replicas)
```

Comment:
I don't think we should consider a node ready to transition to decommissioned state unless all blocks it hosts have 3 replicas elsewhere. Having just one seems risky as a decommissioned node may never come back and the 1 remaining replica can be lost. In HDFS, assuming replication factor 3, decommissioning completes when all blocks have reached 4 replicas - 3 on other nodes and 1 on the decommissioning node. This also means that the decommissioning node can participate as a source to replicate the blocks from until all replication has completed.

This mean that when decommissioning is complete, the cluster should be healthy with no under replicated containers.

Also in a edge case, consider a 3 node cluster - in theory you could decommission 2 nodes and then the cluster can never repair itself as all blocks would drop to 1 replica. In HDFS, decommission would never complete in this case, but based on the "one replica on a healthy node" statement above, it would complete in Ozone.

We also have to consider block placement policy, which probably means each block needs to be on at least 2 racks or node groups (I am not aware how this works in Ozone right now).

Reply:
I tried to make more clear this stop condition in the next version of the document. I agree with you that we need 3 replicas on other nodes from all the containers but what about other nodes in maintenance mode?

This line would like to define that we need 3 replicas (defined in the next line) *AND* at least one of the replicas should be server by a HEALTHY node.

So the stop condition for decommissioning should be:
```
(healthy >= 1) && (healthy + maintenance >= 3)
```

(for maintenance mode the second part can be ignored)

Reply:
And as you suggested in an other comment both 1 and 3 can be a configurable variable.

Reply:
Assuming maintenance.min.replicas of 1, then I think the above logic is correct.

You have 2 nodes in maintenance.
1 node hosting the 3rd replica (healthy)

Decommission this healthy node - it will need to make 1 more copy of the block leaving 2 in maintenance and 1 new healthy one.